Don't cache and don't forward hop-by-hop headers

ikoveshnikov opened the pull request:
Fix issue #409 

- Parse Connection header during parsing HTTP message;
- Store raw headers listed in Connection header in temporary array allocated in message's pool: every raw header is marked as hop-by-hop if it's name present in that array. When connection header parsed in the first time, parser checks already parsed raw headers and mark them as hop-by-hop headers if needed. If Connection header does not contain any raw headers names no extra allocations or string comparison happen;
- Special headers have their own marking procedure which does not involve string comparisons. Some of the special headers are always hop-by-hop (Connection, we also mark Server in responses as hop-by-hop) and using generic comparison like for Raw headers will ruin performance;
- Unit tests for responses and requests with hop-by-hop marking checks; 
- little optimisation of __hdr_lookup(): reduce string comparisons.

*TBD:* 
When response contains `Connection: keep-alive`, `Keep-alive` header will be marked as hop-by-hop header and will not be saved in cache (that was the requirement in #409 ). RFC 7230 6.1 also say:
```
 A proxy or gateway MUST parse a received Connection header field before
   a message is forwarded and, for each connection-option in this field,
   remove any header field(s) from the message with the same name as the
   connection-option, and then remove the Connection header field itself
   (or replace it with the intermediary's own connection options for the
   forwarded message).
```
Currently we have no realised way to set default `Keep-Alive` header (not found in issues but here is corresponding TODO: https://github.com/tempesta-tech/tempesta/blob/master/tempesta_fw/http.c#L639 ). So question is: delete `Keep-Alive` header in forwarded messages or not? But messages served from cache will have no such header. Currently i remove it from both forwarded messages: response and request, disabling that is one-line patch.

krizhanovsky commented on tempesta_fw/http_parser.c:
```diff
@@ -457,6 +457,134 @@ parse_int_hex(unsigned char *data, size_t len, unsigned long *acc)
+	if (!ht)
+		return;
+
+	for (i = 0; i < ht->off; ++i) {
```
Does RFC allow setting headers, mentioned in Connection, before Connection header? The code is dangerous, just consider the example:

        GET / HTTP/1.1\r\n
        Connection: keep-alive, Host, User-Agent\r\n
        Host: foo.com\r\n
        User-Agent: malicious client\r\n
        \r\n

So server will receive malformed request.

We should run the loop at least from TFW_HTTP_HDR_RAW. Probably there is some limitation which headers can be mentioned in Connection header... The other way (I think the safest one) is to remove Keep-Alive only and ignore any other mentioned headers in Connection. However, we still use current mechanism for the headers removal such that we'll be able to remove other headers which are known to be mentioned in Connection. I argue against blind headers removal.

ikoveshnikov replied:
1. RFC does not have any restrictions where headers listed in Connection header must be placed. So both variants are ok: before or after Connection header. More over i tried to set up custom hop-by-hop headers in ngingx and apache, both allow setting custom headers before and after the header. Here example from RFC 6455 with hop-by-hop header before Connection header:
```
GET /chat HTTP/1.1
Host: server.example.com
Upgrade: websocket
Connection: Upgrade
Sec-WebSocket-Key: dGhlIHNhbXBsZSBub25jZQ==
Origin: http://example.com
Sec-WebSocket-Protocol: chat, superchat
Sec-WebSocket-Version: 13
```

2. Table you point to stores names for only raw headers. I agree, it could be confusing, but i used data structure which suit me perfectly. In situation you show user-agent and every other spec header listed in connection header (except keep-alive) will not be removed

Obsoleted now, in new revision this array is statically allocated

krizhanovsky commented on tempesta_fw/http_parser.c:
```diff
@@ -457,6 +457,134 @@ parse_int_hex(unsigned char *data, size_t len, unsigned long *acc)
+	TfwStr *hdr, *append;
+	int r;
+	TfwHttpHdrTbl *ht = hm->parser.hbh_parser.raw;
+	if (!ht) {
```
This isn't mentioned in our Coding Style, but [FreeBSD KNF](https://www.freebsd.org/cgi/man.cgi?query=style&sektion=9) requires split variables definitions and code. It's bit hard to read when code immediately follows variable definitions immediately.

krizhanovsky commented on tempesta_fw/http_parser.c:
```diff
@@ -457,6 +457,134 @@ parse_int_hex(unsigned char *data, size_t len, unsigned long *acc)
+}
+
+/**
+ * Add header name listed in Connection header to rable of raw headers.
```
rable -> table

krizhanovsky commented on tempesta_fw/http_parser.c:
```diff
@@ -457,6 +457,134 @@ parse_int_hex(unsigned char *data, size_t len, unsigned long *acc)
+		if (hid == ht->size)
+			if (__tfw_http_msg_grow_hdr_tbl(&hm->parser.hbh_parser.raw,
+							hm->pool))
+				return -ENOMEM;
```
This is true and powerful headers parser. Do we really allow that Connection mentioning more than 16 header? I believe it has sense to block the header. Meantime it will make the code faster and easier and that's good.

krizhanovsky commented on tempesta_fw/http_parser.c:
```diff
@@ -917,8 +1069,24 @@ __parse_connection(TfwHttpMsg *hm, unsigned char *data, size_t len)
 	__FSM_STATE(I_ConnOther) {
-		__FSM_I_MATCH_MOVE(token, I_ConnOther);
+		/* Unwind __FSM_I_MATCH_MOVE(token, I_ConnOther): */
+		__fsm_n = __data_remain(p);
```
There is MATCH_MOVE macroses which embody the code.

krizhanovsky commented on tempesta_fw/http_parser.c:
```diff
@@ -2079,6 +2318,17 @@ tfw_http_parse_req(void *req_data, unsigned char *data, size_t len)
+		case 'k':
+			if (likely(__data_available(p, 11)
+				   && C4_INT_LCM(p + 1, 'e', 'e', 'p', '-')
+				   && C4_INT_LCM(p + 5, 'a', 'l', 'i', 'v')
```
The discussed bug to fix

krizhanovsky commented on tempesta_fw/t/unit/test_http_parser.c:
```diff
@@ -877,6 +877,219 @@ TEST(http_parser, cookie)
 }
 
+TEST(http_parser, req_hop_by_hop)
+{
```
Please add test case described above for "Connection: Host\r\n' to mangled message.

krizhanovsky commented on tls/ttls.c:
```diff
@@ -128,7 +128,7 @@ module_exit(ttls_exit);
 MODULE_AUTHOR("Tempesta Technologies");
 MODULE_VERSION("2.3.0");
-MODULE_LICENSE("GPLv2");
+MODULE_LICENSE("GPL v2");
```
Doesn't it taints kernel? The change is fine if not.

krizhanovsky commented:
Please address the comment https://github.com/tempesta-tech/tempesta/issues/409#issuecomment-258436328 or move the requirement to #634

ikoveshnikov commented:
Changes since last review:
- rebased to current master
- fixed error in current master with Transfer-Encoding header: the header now is SPEC header but it missed in `__http_msg_hdr_val()`
- Connection header is limited to have less than 16 connection tokens for raw headers. Otherwise request or response will be filtered 
- Drop message if some of connection tokens is end-to-end headers used by tempesta and parsed by TFW_HTTP_PARSE_RAWHDR_VAL macro. Comparing of connection tokens with all headers defined in RFC 7230, 7231 as end-to-end will lead to a lot of string comparisons inside of parsing of Connection header 
- Unit tests updated to have more test cases for hop-by-hop headers marking

The feature have no special requirements for #634 . (Headers listed there are not hop-by-hop)

keshonok commented on tempesta_fw/http_msg.c:
```diff
@@ -573,12 +577,62 @@ tfw_http_msg_hdr_xfrm(TfwHttpMsg *hm, char *name, size_t n_len,
+
+	h = &ht->tbl[hid];
+	TFW_STR_FOR_EACH_DUP(d, h, end)
+		d->flags |= TFW_STR_HBH_HDR;
```
Do we really need to mark each duplicate as HBH? I don't think so, but maybe there's a case to prove otherwise?

keshonok commented on tempesta_fw/http_msg.c:
```diff
@@ -573,12 +577,62 @@ tfw_http_msg_hdr_xfrm(TfwHttpMsg *hm, char *name, size_t n_len,
+ * name only is enough in @hdr.
+ */
+void
+tfw_http_msg_mark_hbh_hdr(TfwHttpMsg *hm, TfwStr *hdr)
```
I believe that this function (under a better name) belongs in `http_parser.c`.
If I understand correctly, the reason for placing it here was that `__hdr_lookup()` was located here, and it's a local (static) function to this sub-module. In my opinion though it should be the other way around: `__hdr_lookup()` should be made public (and given a proper name for that) and used from `http_parser.c`

keshonok commented on tempesta_fw/http_parser.c:
```diff
@@ -462,30 +462,51 @@ parse_int_hex(unsigned char *data, size_t len, unsigned long *acc)
+{
+	TfwHttpHbhHdrs *hbh_hdrs = &req->parser.hbh_parser;
+	/* Connection is hop-by-hop header by RFC 7230 6.1 */
+	hbh_hdrs->spec |= 0x1 << TFW_HTTP_HDR_CONNECTION;
```
As this should be performed only once per message, shouldn't this be a direct assignment, not an OR operation?

keshonok commented on tempesta_fw/http_parser.c:
```diff
@@ -462,30 +462,51 @@ parse_int_hex(unsigned char *data, size_t len, unsigned long *acc)
 	hbh_hdrs->spec |= 0x1 << TFW_HTTP_HDR_CONNECTION;
-	if (!req)
-		hbh_hdrs->spec |= 0x1 << TFW_HTTP_HDR_SERVER;
+	hbh_hdrs->spec |= 0x1 << TFW_HTTP_HDR_SERVER;
```
Please see the comment above regarding direct assignment vs OR operation.

keshonok commented on tempesta_fw/http_parser.c:
```diff
@@ -1348,6 +1348,7 @@ __parse_transfer_encoding(TfwHttpMsg *hm, unsigned char *data, size_t len)
 /* Main (parent) HTTP request processing states. */
 enum {
 	Req_0,
+	Req_CRLF,
```
In my opinion, this may appear too generic. Something like **Req_0_CRLF** may visibly clarify that this is one of the initial states.

ikoveshnikov commented on tempesta_fw/http_parser.c:
```diff
@@ -2089,16 +2090,20 @@ tfw_http_parse_req(void *req_data, unsigned char *data, size_t len)
+	/* Parser internal initilizers, must be called once per message. */
 	__FSM_STATE(Req_0) {
+		__hbh_parser_init_req(req);
+		/* fall through */
```
I have checked out all the places in code, where _fall through_ used instead of `FSM_MOVE` directives. No errors there and  _fall through_ can be replaced with `__FSM_JMP`. Since desired state is the next one _fall through_ is used.

`__FSM_STATE` saves current parser state to `__fsm_const_state`: https://github.com/tempesta-tech/tempesta/blob/master/tempesta_fw/http_parser.c#L120. In all the places in parser _fall through_ is used only when we do not move our current position in message and we want to work with the same data in next stage or stages. 

Any flavour of `FSM_MOVE*` is not an option here since `parser->to_go` will be set to 0 for situation you are describing ( https://github.com/tempesta-tech/tempesta/blob/master/tempesta_fw/http_parser.c#L137 ). And next stage will take *next* character, not the same one.

krizhanovsky commented on tempesta_fw/http_parser.c:
```diff
@@ -887,8 +1081,38 @@ __FSM_STATE(RGen_OWS) {							\
+		parser->hbh_parser.spec |= 0x1 << hid;			\
+		if (!TFW_STR_EMPTY(&msg->h_tbl->tbl[hid]))		\
+			msg->h_tbl->tbl[hid].flags |= TFW_STR_HBH_HDR;	\
+	})
```
The macro is used in only one place, so the code would be more readable if you place the macro as inline code at the appropriate place.

krizhanovsky commented on tempesta_fw/http_parser.c:
```diff
@@ -887,8 +1081,38 @@ __FSM_STATE(RGen_OWS) {							\
+			msg->h_tbl->tbl[hid].flags |= TFW_STR_HBH_HDR;	\
+	})
+
+#define TRY_SPEC_HBH_HDR(name, hid) TRY_SPEC_HBH_HDR_LAMBDA(name, hid, {})
```
The macro is not used at all.

krizhanovsky commented on tempesta_fw/http_parser.c:
```diff
@@ -1781,9 +2101,15 @@ tfw_http_parse_req(void *req_data, unsigned char *data, size_t len)
 
+	/* Parser internal initilizers, must be called once per message. */
 	__FSM_STATE(Req_0) {
+		__hbh_parser_init_req(req);
```
Why did you create a new parser state for the initialization routine? It seems you can call __hbh_parser_init_*() functions in `tfw_http_msg_alloc()` where base parser is initialized.

krizhanovsky commented on tempesta_fw/http_parser.c:
```diff
@@ -3264,9 +3558,15 @@ tfw_http_parse_resp(void *resp_data, unsigned char *data, size_t len)
 
+	/* Parser internal initilizers, must be called once per message. */
 	__FSM_STATE(Resp_0) {
+		__hbh_parser_init_resp(resp);
```
The same as for request

krizhanovsky commented on tempesta_fw/t/unit/test_http_parser.c:
```diff
@@ -950,6 +982,309 @@ TEST(http_parser, cookie)
+
+	/* Connection header lists end-to-end spec headers */
+	FOR_REQ(REQ_HBH_START
+		"Connection: Host, Content-Length, Content-Type, Connection,"
```
I believe the request must be blocked since there is special header in `Connection`
